### PR TITLE
[Debug] Trigger a deprecation when using an internal class/trait/interface

### PR DESCRIPTION
**Pull request by GuilhemN**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no
| Deprecations? | not truly 
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/issues/23800
| License       | MIT
| Doc PR        | 

Trigger a deprecation when the user uses an internal class/trait/interface. 
The deprecation is not triggered when an `@internal` is used in the same vendor.

### CODE REVIEW

**Comment by GuilhemN** [comment]
Fabbot failure unrelated.

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 1]
```diff
@@ -166,10 +167,14 @@ public function loadClass($class)
             }
 
             $parent = get_parent_class($class);
+            $doc = $refl->getDocComment();
+            if (preg_match('#\n \* @internal(?: .+?)?\r?\n \*(?: @|/$)#s', $doc, $notice)) {
+                self::$internal[$name] = true;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: we should keep an reuse the notice in the same way as we do for "final"

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 2]
```diff
@@ -203,50 +208,73 @@ public function loadClass($class)
...
+                    case 'Symfony\Bridge\\':
+                    case 'Symfony\Bundle\\':
+                    case 'Symfony\Component\\':
+                        $ns = 'Symfony\\';
+                        $len = strlen($ns);
+                        break;
+                }
+            }
+
+            foreach (call_user_func(function () use ($name, $parent) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: can't we make this closure into a method? that'd ease perf and reading

Comment (GuilhemN) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: I just removed it in favor of an `array_merge` call

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 1 above)
Comment (GuilhemN) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: Done. I updated the code base to have consistent messages.

File: src/Symfony/Component/Validator/Context/ExecutionContext.php [hunk 3]
```diff
@@ -30,8 +30,7 @@
  *
  * @see ExecutionContextInterface
  *
- * @internal You should not instantiate or use this class. Code against
- *           {@link ExecutionContextInterface} instead.
+ * @internal since version 3.5. Code against ExecutionContextInterface instead.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Context/ExecutionContext.php]: 2.5?

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Validator/Context/ExecutionContext.php (hunk 3 above)
Comment (GuilhemN) [inline src/Symfony/Component/Validator/Context/ExecutionContext.php]: Oops, indeed

**Comment by nicolas-grekas** [comment]
Thank you @GuilhemN.

**Comment by nicolas-grekas** [comment]
I merged quickly so you can continue on #23816 (reviewers, if you had any comment here, please comment on this other PR)

**Pull request closed without merging**